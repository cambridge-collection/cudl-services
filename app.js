//Config
config = require('../config/base.js');
users = require('../config/users.js');

//Modules
var express = require('express');
var path = require('path');
var favicon = require('static-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
var mysql = require('mysql');
passport = require('passport')
var strategy = require('passport-localapikey').Strategy;
var fs = require('fs-extra');

//cache directories
fs.ensureDir(config.cacheDir);
fs.ensureDir(config.cacheDir+'/transcriptions');
fs.ensureDir(config.cacheDir+'/translations');

//Routes
var routes = require('./routes/index.js');
var metadata = require('./routes/metadata.js');
var transcription = require('./routes/transcription.js');
var translation = require('./routes/translation.js');
var membership = require('./routes/membership.js');
var embedded = require('./routes/embedded.js');
var iiif = require('./routes/iiif.js');
var app = express();

//MySQL Connection
connection = mysql.createPool({
	connectionLimit	: 10,
        host     : config.mysqlHost,
        user     : config.mysqlUser,
        password : config.mysqlPass,
	database : config.mysqlData,
});


function findByApiKey(apikey, fn) {
    if (apikey in users) {
      return fn(null, users[apikey]);
    }
  return fn(null, null);
}


passport.use(new strategy(
  function(apikey, done) {
    process.nextTick(function () {
      findByApiKey(apikey, function(err, user) {
        if (err) { return done(err); }
        if (!user) { return done(null, false, { message: 'Unknown apikey : ' + apikey }); }
        return done(null, user);
      })
    });
  }
));


// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(favicon(path.join(__dirname, '/public/images/brand/favicon.ico')));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded());
app.use(cookieParser());
app.use(passport.initialize());
app.use(express.static(path.join(__dirname, 'public')));

//app.use('/', routes);
app.use('/v1/metadata', metadata);
app.use('/v1/transcription',transcription);
app.use('/v1/translation', translation);
app.use('/v1/rdb/membership', membership);
app.use('/v1/embedded', embedded);
app.use('/v1/iiif', iiif);

/// catch 404 and forward to error handler
app.use(function(req, res, next) {
    var err = new Error('Not Found');
    err.status = 404;
    next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
    app.use(function(err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: {}
    });
});

console.log('hmmm');
module.exports = app;
