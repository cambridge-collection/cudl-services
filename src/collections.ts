import {BasePostgresDAO} from './db';
import {Resource} from './resources';

export interface Collection {
  title: string;
  collectionID: string;
  collectionOrder: number;
}

export type GetItemCollections = (itemID: string) => Promise<Collection[]>;

export interface CollectionDAO extends Resource {
  getItemCollections: GetItemCollections;
}

export class PostgresCollectionDAO
  extends BasePostgresDAO
  implements CollectionDAO
{
  async getItemCollections(itemID: string): Promise<Collection[]> {
    const sql = `\
WITH RECURSIVE collection_membership(collectionid, title, collectionorder) AS (
  SELECT collections.collectionid, title, collectionorder, parentcollectionid
  FROM collections
         JOIN itemsincollection on collections.collectionid = itemsincollection.collectionid
  WHERE itemsincollection.itemid = $1::text AND itemsincollection.visible
  UNION
  SELECT collections.collectionid, collections.title, collections.collectionorder, collections.parentcollectionid
  FROM collections,
       collection_membership
  WHERE collections.collectionid = collection_membership.parentcollectionid
)
SELECT *
FROM collection_membership;`;
    return (await this.db.query(sql, [itemID])).rows.map(row => ({
      title: row.title,
      collectionID: row.collectionid,
      collectionOrder: Number(row.collectionorder),
    }));
  }
}
