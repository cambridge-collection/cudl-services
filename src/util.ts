import escapeStringRegexp from 'escape-string-regexp';
import { Request } from 'express';
import url from 'url';
import * as util from 'util';
import { ValueError } from './errors';

const CUDL_HOST = 'cudl.lib.cam.ac.uk';
const CUDL_HOST_REGEX = new RegExp(
  '(?:^|\\.)' + escapeStringRegexp(CUDL_HOST) + '$'
);

export const CORS_HEADERS = Object.freeze({
  'Access-Control-Allow-Origin': '*',
});

export function isExternalCorsRequest(req: Request) {
  const origin = req.header('origin');
  if (!origin) {
    return false;
  }

  const host = url.parse(origin).hostname;

  // If we have an origin header and it's not cudl, then it's an external cors
  // request.
  return typeof host === 'string' && !CUDL_HOST_REGEX.test(host);
}

export function isSimplePathSegment(value: string): boolean {
  return /^[\w -]+$/.test(value);
}

/**
 * A type guard which narrows strings to members of string-valued enums.
 *
 * @param _enum The enum type
 * @param val The value to narrow
 */
export function isEnumMember<E extends string>(
  _enum: { [k in string]: E },
  val: string
): val is E {
  const members = Object.values(_enum) as unknown[];
  return members.includes(val);
}

export function requireRequestParam(req: Request, param: string): string {
  const value = req.params[param] as typeof req.params[string] | undefined;
  if (typeof value !== 'string') {
    throw new Error(`Request has no value for param ${util.inspect(param)}`);
  }
  return value;
}

export function requireRequestParams<T extends string>(
  req: Request,
  ...params: T[]
): { [key in T]: string } {
  const obj = {} as { [key in T]: string };
  for (const param of params) {
    obj[param] = requireRequestParam(req, param);
  }
  return obj;
}

export function factory<A, B>(
  constructor: new (...args: A[]) => B
): (...args: A[]) => B {
  return (...args: A[]) => new constructor(...args);
}

/** The type of the only argument of a 1-ary constructor. */
export type UnaryConstructorArg<
  T extends new (arg: unknown) => unknown
> = ConstructorParameters<T>[0];

export type ComparePrimitive = string | number | boolean | ComparePrimitive[];
export type CompareModifier = [Direction, CompareValue];
export type CompareValue = ComparePrimitive | CompareModifier | CompareValue[];

export function compare(a: CompareValue, b: CompareValue): number {
  if (typeof a !== typeof b) {
    throw new Error(
      `compare() not supported between ${typeof a} and ${typeof b}`
    );
  }

  if (!Array.isArray(a)) {
    return a === b ? 0 : a < b ? -1 : 1;
  }

  // Directional compare modifier
  if ((a.length > 0 && a[0] === asc) || a[0] === desc) {
    b = b as CompareModifier | CompareValue[];
    if (a.length !== 2) {
      throw new Error(`Invalid compare direction modifier: modifiers must be a pair of [direction, value], \
got: ${util.inspect(a)}`);
    }
    if (b.length !== 2 || b[0] !== a[0]) {
      throw new Error(
        `Invalid compare direction modifier: modifiers on both comparison values must use the same direction`
      );
    }

    return compare(a[1], b[1]) * (a[0] === desc ? -1 : 1);
  }

  a = a as CompareValue[];
  b = b as CompareValue[];

  for (let i = 0; i < Math.min(a.length, b.length); ++i) {
    const order = compare(a[i], b[i]);
    if (order !== 0) {
      return order;
    }
  }
  // shorter array is less
  return a.length - b.length;
}

const asc: unique symbol = Symbol('compare.asc');
const desc: unique symbol = Symbol('compare.desc');
compare.asc = asc;
compare.desc = desc;
type Direction = typeof asc | typeof desc;

export function sorted<T extends ComparePrimitive>(items: Iterable<T>): T[];
export function sorted<T>(
  items: Iterable<T>,
  key: (value: T) => CompareValue
): T[];
export function sorted<T>(
  items: Iterable<T>,
  key?: (value: T) => CompareValue
): T[] {
  if (key === undefined) {
    const sorted = Array.from(
      (items as Iterable<unknown>) as Iterable<ComparePrimitive>
    );
    sorted.sort((a, b) => compare(a, b));
    return (sorted as unknown[]) as T[];
  }

  const indexedItems = index(items);
  const keys = indexedItems.map(([_, value]) => key(value));
  indexedItems.sort(([ia], [ib]) => compare(keys[ia], keys[ib]));

  // Remove indexes in place
  const sorted = (indexedItems as unknown) as T[];
  for (let i = 0; i < indexedItems.length; ++i) {
    sorted[i] = indexedItems[i][1];
  }

  return sorted;
}

function index<T>(items: Iterable<T>): Array<[number, T]> {
  const index: Array<[number, T]> = [];
  let i = 0;
  for (const item of items) {
    index[i] = [i, item];
    i++;
  }
  return index;
}
