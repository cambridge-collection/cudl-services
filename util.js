var url = require("url");
var regexpQuote = require("regexp-quote");


var CUDL_HOST = "cudl.lib.cam.ac.uk";
var CUDL_HOST_REGEX = new RegExp("(?:^|\\.)" + regexpQuote(CUDL_HOST) + "$");

var CORS_HEADERS = module.exports.CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*"
};


module.exports.isExternalCorsRequest = function isExternalCorsRequest(req) {
    var origin = req.header("origin") || "";
    if(!origin) { return false; }

    var host = url.parse(origin).hostname;

    // If we have an origin header and it's not cudl, then it's an external cors
    // request.
    return !CUDL_HOST_REGEX.test(host);
};
