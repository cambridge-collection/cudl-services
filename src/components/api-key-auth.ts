import {fnComponent, Component} from '../app';
import passport from 'passport';
import express from 'express';

const AccessTokenStrategy = require('passport-accesstoken').Strategy;

export interface Users<U = User> {
  [apiKey: string]: U;
}

export interface User {
  username: string;
  email: string;
}

export function passportApiKeyAuthenticationComponent(users: Users) {
  return fnComponent(express => {
    getPassport(express)
      .use(
        new AccessTokenStrategy(
          (token: string, done: (err: unknown, user: unknown) => void) => {
            process.nextTick(() => {
              const user = findByApiKey(users, token);
              return done(null, user || false);
            });
          }
        )
      )
      .serializeUser((user, done) => {
        done(null, user);
      });
  });
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}

export const passportInitialiseComponent: Component = fnComponent(express => {
  const localPassport = new passport.Authenticator();
  express.set('passport', localPassport);
  express.use(localPassport.initialize());
});

export function getPassport(app: express.Application): passport.Authenticator {
  const localPassport = app.get('passport');
  if (localPassport !== undefined && typeof localPassport === 'object') {
    return localPassport as passport.Authenticator;
  }
  throw new Error(
    'express app has no passport Authenticator on the "passport" setting. Has the passportInitialiseComponent been registered?'
  );
}
