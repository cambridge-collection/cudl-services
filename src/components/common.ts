import {Component, MiddlewareComponent, SettingsComponent} from '../app';
import path from 'path';
import bodyParser from 'body-parser';
import express from 'express';
import {
  passportApiKeyAuthenticationComponent,
  passportInitialiseComponent,
  Users,
} from './api-key-auth';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');

/**
 * Components to register ahead of regular request handling middleware.
 *
 * @param options
 */
export function leadingComponents(options?: {apiKeys?: Users}): Component[] {
  options = options || {};

  const authMiddleware =
    options.apiKeys === undefined
      ? []
      : [
          passportInitialiseComponent,
          passportApiKeyAuthenticationComponent(options.apiKeys),
        ];

  return [
    // Disable parsing of query strings into nested objects
    new SettingsComponent({'query parser': 'simple'}),
    new MiddlewareComponent({
      handler: [
        favicon(
          path.resolve(__dirname, '../../public/images/brand/favicon.ico')
        ),
        logger('dev'),
        bodyParser.json(),
        bodyParser.urlencoded({extended: false}),
        cookieParser(),
        express.static(path.resolve(__dirname, '../../public')),
        stripTrailingSlashes,
      ],
    }),
    ...authMiddleware,
  ];
}

/**
 * Middleware to redirect trailing slashes to same URL without trailing slash
 */
const stripTrailingSlashes: express.Handler = (req, res, next) => {
  if (req.url.substr(-1) === '/' && req.url.length > 1) {
    res.redirect(301, req.url.slice(0, -1));
  } else {
    next();
  }
};
