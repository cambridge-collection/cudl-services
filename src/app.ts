import { XSLTExecutor } from '@lib.cam/xslt-nailgun';
import bodyParser from 'body-parser';
import Debugger from 'debug';
import express, { NextFunction, Request, Response } from 'express';
import passport from 'passport';
import path from 'path';
import { CollectionDAO, PostgresCollectionDAO } from './collections';
import { Config, User, Users } from './config';

import { DAOPool, PostgresDatabasePool } from './db';
import {
  CUDLMetadataRepository,
  DefaultCUDLMetadataRepository,
  LegacyDarwinMetadataRepository,
} from './metadata';
import { BaseResource, ExternalResources, using } from './resources';
import * as darwin from './routes/darwin';
import * as membership from './routes/membership';
import * as metadata from './routes/metadata';
import * as similarity from './routes/similarity';
import * as tags from './routes/tags';
import { PostgresTagsDAO, TagsDAO } from './routes/tags-impl';
import * as transcription from './routes/transcription';
import * as translation from './routes/translation';
import { DefaultXTF, XTF } from './xtf';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');
const Strategy = require('passport-accesstoken').Strategy;

const debug = Debugger('cudl-services');

// FIXME: move this to server?
//cache directories
// fs.ensureDirSync(config.cacheDir);
// fs.ensureDirSync(config.cacheDir+'/transcriptions');
// fs.ensureDirSync(config.cacheDir+'/translations');

export interface AppOptions {
  users: Users;
  metadataRepository: CUDLMetadataRepository;
  legacyDarwinMetadataRepository: LegacyDarwinMetadataRepository;
  collectionsDAOPool: DAOPool<CollectionDAO>;
  tagsDAOPool: DAOPool<TagsDAO>;
  darwinXtfUrl: string;
  xtf: XTF;
}

export class App extends BaseResource {
  readonly options: AppOptions;
  readonly expressApp: express.Application;
  protected readonly xsltExecutor: XSLTExecutor;

  constructor(options: AppOptions) {
    super();
    this.options = Object.freeze({ ...options });
    this.xsltExecutor = XSLTExecutor.getInstance();
    this.expressApp = this.createExpressApp();
  }

  static fromConfig(config: Config): ExternalResources<App> {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: new DefaultCUDLMetadataRepository(config.dataDir),
        legacyDarwinMetadataRepository: new LegacyDarwinMetadataRepository(
          config.legacyDcpDataDir
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
      }),
      [dbPool]
    );
  }

  private createExpressApp(): express.Application {
    const app = express();

    passport.use(
      new Strategy((token: string, done: (err: any, user: any) => void) => {
        process.nextTick(() => {
          const user = findByApiKey(this.options.users, token);
          return done(null, user || false);
        });
      })
    );

    // view engine setup
    app.set('views', path.resolve(__dirname, '../views'));
    app.set('view engine', 'pug');

    app.use(
      favicon(path.resolve(__dirname, '../public/images/brand/favicon.ico'))
    );
    app.use(logger('dev'));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(cookieParser());
    app.use(passport.initialize());
    app.use(express.static(path.resolve(__dirname, '../public')));

    // Middleware to redirect trailing slashes to same URL without trailing slash
    app.use((req, res, next) => {
      if (req.url.substr(-1) === '/' && req.url.length > 1) {
        res.redirect(301, req.url.slice(0, -1));
      } else {
        next();
      }
    });

    app.use(
      '/v1/metadata',
      metadata.getRoutes({
        metadataRepository: this.options.metadataRepository,
      })
    );

    app.use(
      '/v1/rdb/membership',
      membership.getRoutes({
        getItemCollections: async (itemID: string) =>
          using(this.options.collectionsDAOPool.getInstance(), dao =>
            dao.getItemCollections(itemID)
          ),
      })
    );

    app.use('/v1/tags', tags.getRoutes({ daoPool: this.options.tagsDAOPool }));

    app.use(
      '/v1/transcription',
      transcription.getRoutes({
        metadataRepository: this.options.metadataRepository,
        legacyDarwinMetadataRepository: this.options
          .legacyDarwinMetadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/translation',
      translation.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/xtf/similarity',
      similarity.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xtf: this.options.xtf,
      })
    );

    app.use(
      '/v1/darwin',
      passport.authenticate('token', { session: false }),
      darwin.getRoutes({ darwinXtfUrl: this.options.darwinXtfUrl })
    );

    // 404 if no route matched
    app.use((req: Request, res: Response, next: NextFunction) => {
      res.status(404).send('Not Found');
    });

    return app;
  }

  async close(): Promise<void> {
    await Promise.all([super.close(), this.xsltExecutor.close()]);
  }
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}
