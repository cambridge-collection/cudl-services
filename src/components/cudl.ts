import {DAOPool} from '../db';
import {Components, MiddlewareComponent} from '../app';
import * as tags from '../routes/tags';
import * as translation from '../routes/translation';
import * as similarity from '../routes/similarity';
import {CollectionDAO} from '../collections';
import {
  CUDLMetadataRepository,
  cudlProvidersForDataStore,
} from '../metadata/cudl';
import {URL} from 'url';
import * as transcription from '../routes/transcription';
import {XSLTExecutor} from '@lib.cam/xslt-nailgun';
import * as metadata from '../routes/metadata';
import {DomainNameMatcher, ExternalCorsRequestMatcher} from '../util';
import {DataStore, ItemJsonMetadataResponseEmitter} from '../metadata';
import * as membership from '../routes/membership';
import {using} from '../resources';

export type TagsOptions = Omit<tags.GetRoutesOptions, 'router'>;
export function tagsComponents(options: TagsOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/tags',
    handler: tags.getRoutes({daoPool: options.daoPool}),
  });
}

export interface TranscriptionOptions {
  metadataRepository: CUDLMetadataRepository;
  teiServiceURL: URL;
  xsltExecutor: XSLTExecutor;
  zacynthiusServiceURL: URL;
}
export function transcriptionComponents(
  options: TranscriptionOptions
): Components {
  return new MiddlewareComponent({
    path: '/v1/transcription',
    handler: transcription.getRoutes({
      metadataRepository: options.metadataRepository,
      teiServiceURL: options.teiServiceURL,
      xsltExecutor: options.xsltExecutor,
      zacynthiusServiceURL: options.zacynthiusServiceURL,
    }),
  });
}

export type TranslationOptions = Omit<translation.GetRoutesOptions, 'router'>;
export function translationComponents(options: TranslationOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/translation',
    handler: translation.getRoutes(options),
  });
}

export interface MetadataOptions {
  cudlDataDataStore: DataStore;
  internalDomainName: string;
}
export function metadataComponents(options: MetadataOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/metadata',
    handler: metadata.getRoutes({
      metadataProviders: new Map(
        Object.entries(cudlProvidersForDataStore(options.cudlDataDataStore))
      ),
      isExternalCorsRequest: ExternalCorsRequestMatcher({
        internalDomains: DomainNameMatcher(options.internalDomainName),
      }),
      metadataEmitters: [ItemJsonMetadataResponseEmitter.instance],
    }),
  });
}

export interface CollectionMembershipOptions {
  collectionsDAOPool: DAOPool<CollectionDAO>;
}
export function collectionMembershipComponents(
  options: CollectionMembershipOptions
): Components {
  return new MiddlewareComponent({
    path: '/v1/rdb/membership',
    handler: membership.getRoutes({
      getItemCollections: async (itemID: string) =>
        using(options.collectionsDAOPool.getInstance(), dao =>
          dao.getItemCollections(itemID)
        ),
    }),
  });
}

export type SimilarityOptions = Omit<similarity.GetRouteOptions, 'router'>;
export function similarityComponents(options: SimilarityOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/xtf/similarity',
    handler: similarity.getRoutes(options),
  });
}
