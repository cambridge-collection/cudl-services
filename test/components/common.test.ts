import request from 'supertest';
import {leadingComponents} from '../../src/components/common';
import {Application, ComponentApp} from '../../src/app';
import express from 'express';
import expressAsyncHandler from 'express-async-handler';
import {EXAMPLE_STATIC_FILES, STATIC_FILES} from '../constants';
import {StatusCodes} from 'http-status-codes';
import {promisify} from 'util';
import fs from 'fs';
import path from 'path';
import {mocked} from 'ts-jest/utils';
import {getPassport} from '../../src/components/api-key-auth';

describe('leading & trailing', () => {
  let componentApp: Application;
  let app: express.Application;

  beforeEach(async () => {
    jest.clearAllMocks();
    componentApp = await ComponentApp.from(leadingComponents());
    app = componentApp.expressApp;
  });

  describe('authentication', () => {
    test('is not registered by default', () => {
      expect(() => getPassport(app)).toThrowErrorMatchingSnapshot();
    });

    test('is registered if apiKeys are provided', async () => {
      const app = (
        await ComponentApp.from(
          leadingComponents({
            apiKeys: {secret: {email: 'foo@example.com', username: 'foo'}},
          })
        )
      ).expressApp;
      app.use(getPassport(app).authenticate('token'));
      app.use((req, res) => res.end());

      expect((await request(app).get('/')).status).toBe(
        StatusCodes.UNAUTHORIZED
      );
      expect(
        (await request(app).get('/').set('x-token', 'secret')).status
      ).toBe(StatusCodes.OK);
    });
  });

  test('favicon', async () => {
    const resp = await request(app).get('/favicon.ico');
    expect(resp.ok).toBeTruthy();
    expect(resp.type).toBe('image/x-icon');
  });

  describe('query parsing', () => {
    test('query strings are parsed with the simple parser', () => {
      expect(app.get('query parser')).toBe('simple');
    });

    test("the simple query parser doesn't produce nested objects", async () => {
      app.get(
        '/',
        expressAsyncHandler(async (req, resp) => {
          expect(req.query.a).toEqual(['1', '3']);
          expect(req.query.b).toEqual('2');
          expect(req.query['foo[bar]']).toEqual('baz');
          expect(req.query.foo).toBeUndefined();
          resp.send();
        })
      );
      await request(app).get('/?a=1&b=2&a=3&foo[bar]=baz');
      expect.assertions(4);
    });
  });

  describe('static files', () => {
    test.each(Object.values(EXAMPLE_STATIC_FILES).map(sf => [sf]))(
      'static file %o is served',
      async (resource: {path: string; type: string}) => {
        const response = await request(app)
          .get(`/${resource.path}`)
          .responseType('blob');
        expect(response.status).toBe(StatusCodes.OK);
        expect(response.type).toBe(resource.type);

        expect(response.body).toEqual(
          await promisify(fs.readFile)(
            path.resolve(STATIC_FILES, resource.path)
          )
        );
      }
    );
  });

  test('json request bodies are parsed', async () => {
    app.post('/json', (req, res) => {
      expect(req.body).toEqual({json: true});
      res.end();
    });
    const resp = await request(app).post('/json').send({json: true});
    expect(resp.ok).toBeTruthy();
    expect.assertions(2);
  });

  test('urlencoded request bodies are parsed', async () => {
    app.post('/urlencoded', (req, res) => {
      expect(req.body).toEqual({a: '1', b: '2', foo: 'bar'});
      res.end();
    });
    const resp = await request(app)
      .post('/urlencoded')
      .type('urlencoded')
      .send('a=1&b=2&foo=bar');
    expect(resp.ok).toBeTruthy();
    expect.assertions(2);
  });

  test('cookies are parsed', async () => {
    app.get('/cookies', (req, res) => {
      expect(req.cookies).toEqual({foo: 'bar', bar: 'baz'});
      res.end();
    });
    const resp = await request(app)
      .get('/cookies')
      .set('Cookie', ['foo=bar', 'bar=baz']);
    expect(resp.ok).toBeTruthy();
    expect.assertions(2);
  });

  test('requests are logged', async () => {
    jest.spyOn(process.stdout, 'write');
    app.get('/foo', (req, res) => {
      res.end();
    });
    const resp = await request(app).get('/foo');
    expect(resp.ok).toBeTruthy();
    const output = mocked(process.stdout.write)
      .mock.calls.map(args => args[0])
      .join('');
    expect(output).toMatch(/GET \/foo .*200/);
  });

  test('trailing slashes on paths are removed via redirect', async () => {
    const resp = await request(app).get('/foo/');
    expect(resp.status).toBe(StatusCodes.MOVED_PERMANENTLY);
    expect(resp.headers.location).toBe('/foo');
  });

  test('unmatched requests 404', async () => {
    expect((await request(app).get('/foo')).status).toBe(StatusCodes.NOT_FOUND);
  });
});
