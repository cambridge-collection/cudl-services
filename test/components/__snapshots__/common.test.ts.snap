// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leading & trailing authentication is not registered by default 1`] = `"express app has no passport Authenticator on the \\"passport\\" setting. Has the passportInitialiseComponent been registered?"`;
